src/gluonts/dataset/parallelized_loader.py

```diff
@@ -346,6 +357,51 @@ def _worker_fn(
     return buf.getvalue()
 
 
+class ShuffleIter(Iterator[DataEntry]):
+    def __init__(
+        self, base_iterator: Iterator[DataEntry], shuffle_buffer_length: int
+    ):
+        self.shuffle_buffer: list = []
+        self.shuffle_buffer_length = shuffle_buffer_length
+        self.base_iterator = base_iterator
+        self.base_iter_finished = False
+
+    def __next__(self) -> DataEntry:
+        # if the buffer is empty, fill the buffer first.
+        # (should only executed in the first round)
+        if not self.shuffle_buffer:
+            self.shuffle_buffer = list(
+                itertools.islice(
+                    self.base_iterator, self.shuffle_buffer_length
+                )
+            )
+        # if buffer still empty, means all elements used,
+        # return a signal of end of iterator
+        if not self.shuffle_buffer:
+            return {}
+        # choose an element at a random index and yield it
+        # and fill it with the next element in the sequential generator
+        idx = random.randint(0, len(self.shuffle_buffer) - 1)
+        next_sample = self.shuffle_buffer[idx]
+
+        # replace the index with the next element in the iterator if the iterator has not finished.
+        # delete the index otherwise.
+        to_replace: DataEntry = next(self.base_iterator, {})
+        if to_replace:
+            self.shuffle_buffer[idx] = to_replace
+        else:
+            del self.shuffle_buffer[idx]
+
+        return next_sample
+
+    def __iter__(self) -> Iterator[DataEntry]:
+        while True:
+            next_sample = next(self)
+            if not next_sample:
+                return
+            yield next_sample
```

Comment:
I think this can simply return `self`, see here: https://docs.python.org/3/library/stdtypes.html#iterator-types

Reply:
Actually I found out that we don't need to implement ```__iter__``` method. Changing ```__next__``` is everything needed.